### PR TITLE
Added note visualization to Arp

### PR DESCRIPTION
**Pull request by jmidt**
The dots are recalculated in the graphics when changing playmode/style or octavemode. The positions are auto-scaled in both x- and y-directions.

### CODE REVIEW

File: src/engines/arps/ARP/screen.cpp [hunk 1]
```diff
@@ -11,55 +15,223 @@ namespace otto::engines::arp {
...
+                           util::local_vector<std::pair<nvg::Point, bool>, 48>& dots_coords) noexcept
+  {
+    // Calculate notes
+    //
+    dots.clear();
+    state.reset();
+    do {
+      dots.push_back(octavemode_func_(state, notes, playmode_func_));
+    } while (state.count != 0);
+    if (playmode_func_ != play_modes::chord) dots.pop_back();
```
Comment (topisani) [inline src/engines/arps/ARP/screen.cpp]:
```suggestion
    if (dots.size() > 1) dots.pop_back();
```
I'd say this logic is more correct

File: src/engines/arps/ARP/screen.cpp [hunk 2]
```diff
@@ -11,55 +15,223 @@ namespace otto::engines::arp {
...
+    float x_step_width = 30;
+
+    // Calculate dots from notes
+    //
+    int num_steps = dots.size();
+    int min = 88;
+    int max = 0;
+    // Find minimum and maximum key values
+    for (auto& s : dots) {
+      // s is a set so is it ordered by note?
```
Comment (topisani) [inline src/engines/arps/ARP/screen.cpp]: yes it is
```suggestion
```

File: src/engines/arps/ARP/screen.hpp [hunk 3]
```diff
@@ -12,50 +15,73 @@ namespace otto::engines::arp {
...
+      case Playmode::updowninc: return "UP/DOWN INC.";
+      case Playmode::downupinc: return "DOWN/UP INC.";
+      case Playmode::manual: return "MANUAL";
+      case Playmode::chord: return "CHORD";
+      case Playmode::random: return "RANDOM";
     };
     OTTO_UNREACHABLE;
   }
 
   inline std::string to_string(OctaveMode om) noexcept
```
Comment (topisani) [inline src/engines/arps/ARP/screen.hpp]: these functions used for formating strings for display should probably be called `display` or `to_display_string`
And they should definitely be declared/defined in the cpp file.

Also, they shouldn't return a `std::string`, but instead a `util::string_ref`

**Review by topisani — CHANGES_REQUESTED** [review]

File: src/engines/arps/ARP/screen.cpp (hunk 1 above)
Comment (jmidt) [reply src/engines/arps/ARP/screen.cpp]: Yeah, that's correct.

File: src/engines/arps/ARP/screen.hpp (hunk 3 above)
Comment (jmidt) [reply src/engines/arps/ARP/screen.hpp]: Gotcha!

File: src/engines/arps/ARP/screen.hpp [hunk 4]
```diff
@@ -7,39 +7,14 @@
...
-      case OctaveMode::octaveup: return "+1";
-      case OctaveMode::doubleoctaveup: return "+2";
-      case OctaveMode::octaveupunison: return "+1 UNISON";
-      case OctaveMode::fifthunison: return "FIFTH";
-      case OctaveMode::octavedownup: return "-1 & +1";
-    };
-    OTTO_UNREACHABLE;
-  }
+  inline util::string_ref display(Playmode pm) noexcept;
+  inline util::string_ref display(OctaveMode om) noexcept;
```
Comment (topisani) [inline src/engines/arps/ARP/screen.hpp]: they arent `inline` anymore

Comment (jmidt) [reply src/engines/arps/ARP/screen.hpp]: Hmm. Was it ever? I don't think switch statements play nice with `inline`? 
But in any case, I guess I should just remove the keyword if I put it in the .cpp file?

**Comment by jmidt** [comment]
I have merged in the recent changes from develop. If there is nothing else, I will merge this in when Travis finishes.

**Review by topisani — APPROVED** [review]

**Pull request merged**